astropy/utils/data.py

```diff
@@ -244,16 +246,49 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
             delete_fds.append(tmp)
             fileobj_new = bz2.BZ2File(tmp.name, mode='rb')
             fileobj_new.read(1)  # need to check that the file is really bzip2
-        except IOError:  # invalid bzip2 file
+        except (IOError) as e:  # invalid bzip2 file
             fileobj.seek(0)
             fileobj_new.close()
+            # raise
         else:
             fileobj_new.seek(0)
             close_fds.append(fileobj_new)
             fileobj = fileobj_new
+    elif signature[:3] == b'\xfd7z':  # xz
+        try:
+            # for Python < 3.3 try backports.lzma; pyliblzma installs as lzma,
+            # but does not support TextIOWrapper
+            if sys.version_info >= (3,3,0):
+                import lzma
+                fileobj_new = lzma.LZMAFile(fileobj, mode='rb')
```

Comment:
This is where things get murky, since there's some inconsistencies as to whether the compressed wrapper closes the underlying file handle upon closing between `gzip` and `bz2` and possibly `lzma` as well.  But you may need to do `close_fds.append(fileobj_new)` here to ensure the wrapper gets closed as well.  Only experimentation/testing will tell, though.